bd-report-writer/src/ffi.rs

```diff
@@ -0,0 +1,569 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+// The C and Friends interface to creating a Report from raw parts.
+//
+// To begin, create a processor using `bdrw_create_buffer_handle()`, holding on
+// to this very special pointer for the remaining operations to append to the
+// report, then dispose of it using `bdrw_dispose_buffer_handle()`.
+
+// Wildcard imports seem reasonable once importing 20+ components of a package
+#[allow(clippy::wildcard_imports)]
+use crate::report::bitdrift_public::fbs::issue_reporting::v_1::*;
+use flatbuffers::{FlatBufferBuilder, ForwardsUOffset, Vector, WIPOffset};
+use std::ffi::{c_char, c_void, CStr};
+use std::ptr::null;
+use std::slice;
+
+pub struct ReportProcessor<'a> {
+  builder: FlatBufferBuilder<'a>,
+  binary_images: Vec<WIPOffset<BinaryImage<'a>>>,
+  system_thread_count: u16,
+  threads: Vec<WIPOffset<Thread<'a>>>,
+  errors: Vec<WIPOffset<Error<'a>>>,
+  report_type: ReportType,
+  sdk: WIPOffset<SDKInfo<'a>>,
+  app: Option<WIPOffset<AppMetrics<'a>>>,
+  device: Option<WIPOffset<DeviceMetrics<'a>>>,
+}
+
+// Type alias for the gross pointers passed around for FFI to improve
+// general readability
+pub type BDProcessorHandle = *mut *const c_void;
```

Comment:
you could do a newtype with repr(transparent) if you want to pretend like you have more type safety

Reply:
we kinda do this with with LoggerId in platform/shared if you want to compare notes